contracts/Lock.sol

```diff
@@ -0,0 +1,81 @@
+pragma solidity ^0.4.24;
+
+import "@aragon/os/contracts/apps/AragonApp.sol";
+import "@aragon/os/contracts/common/IForwarder.sol";
+import "@aragon/os/contracts/lib/token/ERC20.sol";
+
+contract Lock is AragonApp, IForwarder {
+
+    bytes32 constant public CHANGE_DURATION_ROLE = keccak256("CHANGE_DURATION_ROLE");
+    bytes32 constant public CHANGE_AMOUNT_ROLE = keccak256("CHANGE_AMOUNT_ROLE");
+
+    string private constant ERROR_NO_TOKENS_LOCKED = "LOCK_NO_TOKENS_LOCKED";
+    string private constant ERROR_WITHIN_LOCK_DURATION = "LOCK_WITHIN_LOCK_DURATION";
+    string private constant ERROR_CAN_NOT_FORWARD = "VOTING_CAN_NOT_FORWARD";
+
+    event ChangeLockDuration(uint256 newLockDuration);
+    event ChangeLockAmount(uint256 newLockAmount);
+
+    struct TokenLock {
+        bool tokensLocked;
+        uint256 lockedUntil;
+    }
+
+    ERC20 token;
+    uint256 lockDuration;
+    uint256 lockAmount;
+    mapping(address => TokenLock) lockedTokens;
+
+    function initialize(address _token, uint256 _lockDuration, uint256 _lockAmount) public onlyInit {
+        token = ERC20(_token);
+        lockDuration = _lockDuration;
+        lockAmount = _lockAmount;
+
+        initialized();
+    }
+
+    function changeLockDuration(uint256 _lockDuration) public auth(CHANGE_DURATION_ROLE) {
+        lockDuration = _lockDuration;
+        emit ChangeLockDuration(lockDuration);
+    }
+
+    function changeLockAmount(uint256 _lockAmount) public auth(CHANGE_AMOUNT_ROLE) {
+        lockAmount = _lockAmount;
+        emit ChangeLockAmount(lockAmount);
+    }
+
+    function withdrawTokens() public {
+        TokenLock storage tokenLock = lockedTokens[msg.sender];
+        require(tokenLock.tokensLocked, ERROR_NO_TOKENS_LOCKED);
+        require(now > tokenLock.lockedUntil, ERROR_WITHIN_LOCK_DURATION);
+
+        tokenLock.tokensLocked = false;
+        token.transfer(msg.sender, lockAmount);
+    }
+
+    function isForwarder() external pure returns (bool) {
+        return true;
+    }
+
+    function canForward(address sender, bytes evmCallScript) public view returns (bool) {
+        bool tokensLocked = lockedTokens[msg.sender].tokensLocked;
+        bool allowanceAvailable = token.allowance(msg.sender, address(this)) >= lockAmount;
+
+        return tokensLocked || allowanceAvailable;
+    }
+
+    // This requires approve be called before, not sure how to do this on the UI. Perhaps an approve pretransaction will work.
+    function forward(bytes evmCallScript) public {
+        require(canForward(msg.sender, evmCallScript), ERROR_CAN_NOT_FORWARD);
+
+        TokenLock storage tokenLock = lockedTokens[msg.sender];
+        tokenLock.lockedUntil = now + lockDuration;
+
+        if (!tokenLock.tokensLocked) {
```

Comment:
If a person tries to forward multiple actions they should be required to lock additional tokens. If i'm understanding this correctly, they would be able to forward unlimited actions as long as their tokens remained locked?

Reply:
Okay, I think the solution would be to have an array of unlock times and when the user comes to withdraw their deposit it checks the lock times in the array and withdraws as many as possible at once. To prevent OOG issues with large arrays there are 3 options:

1) There needs to be a limit on the size of the array, this limits the number of `forward()`'s an account can execute within a period of time. Once they have withdrawn they can execute `forward()` again.
2) The `withdraw()` function needs to accept a parameter which allows it to withdraw only that many times, potentially requiring the user to execute `withdraw()` multiple times. 
3) Similar to the second but that doesn't require an argument to be passed in, using the Solidity `gasLeft()` function which can end execution when not enough gas is left to process the rest of the array, similarly it will potentially require calling `withdraw()` multiple times. This would require investigation into how much gas each withdraw iteration takes and is a bit risky as there's a chance that the `transferFrom()` method a token uses may diverge from the ERC20 standard and have a bigger gas requirement than is unexpected.

Reply:
In hindsight, with what I mentioned in suggestion 3 where `transferFrom()` could become a more costly operation, we probably want to use 2. However, we could have 2 withdraw() functions. `withdraw()` which attempts to withdraw all of the arrays locks that are available and `withdraw(numberOfLocks)` which will withdraw up to the specified number of locks, used when `withdraw()` fails. The UI could display and call the appropriate function depending on which will pass.